src/main/scala/viper/silver/plugin/standard/inline/InlineRewrite.scala

```diff
@@ -0,0 +1,109 @@
+          if getPredicateBody(pred, program).nonEmpty => pred.predicateName
+      }
+    methods.foldLeft(Seq[String]())(_ ++ expandablePredicatesFor(_))
+  }
+
+  def inlinePredicates(
+    method: Method,
+    program: Program
+  ): Method = {
+    val expandedPres = method.pres.map { pre =>
+      expandPredicate(pre, program).fold(pre)(expandedPred => expandedPred)
+    }
+    //    val expandedPosts = method.posts.map { post =>
+    //      expandPredicate(post, program).fold(post)(expandedPred => expandedPred)
+    //    }
+    method.copy(name = method.name,
+    formalArgs = method.formalArgs,
+    formalReturns = method.formalReturns,
+    pres = expandedPres,
+    posts = method.posts,
+    body = method.body
+    )(pos = method.pos, info = method.info, errT = method.errT)
+  }
+
+  def removeUnfoldFold(method: Method, expandablePredIds: Seq[String]): Method = {
+    val rewrittenBody = method.body.map { body =>
+      val bodyWithRemovedUnfolds = removeUnfolds(body.ss, expandablePredIds)
+      body.copy(ss = bodyWithRemovedUnfolds, scopedDecls = body.scopedDecls)(
+        pos = body.pos, info = body.info, errT = body.errT
+      )
+      val bodyWithRemovedUnfoldFolds = removeFolds(bodyWithRemovedUnfolds, expandablePredIds)
+      body.copy(ss = bodyWithRemovedUnfoldFolds, scopedDecls = body.scopedDecls)(
+        pos = body.pos, info = body.info, errT = body.errT
+      )
+    }
+    method.copy(name = method.name,
+    formalArgs = method.formalArgs,
+    formalReturns = method.formalReturns,
+    pres = method.pres,
+    posts = method.posts,
+    body = rewrittenBody
+    )(pos = method.pos, info = method.info, errT = method.errT)
+  }
+
+  private[this] def expandPredicate(expr: Exp, program: Program): Option[Exp] =
+  expr match {
+      case PredicateAccessPredicate(pred, _) =>
+        getPredicateBody(pred, program)
+      case _ => None
+    }
+
+  private[this] def getPredicateBody(pred: PredicateAccess, program: Program): Option[Exp] = {
+    def extractLocalVarIds(predicate: PredicateAccess): Set[String] =
+    predicate.args.map {
+        case LocalVar(name, _) => name
+        case _ => ""
+      }.filter(_.nonEmpty).toSet
+
+    val args = extractLocalVarIds(pred)
+    pred.predicateBody(program, args)
+  }
+
+
+  private[this] def removeUnfolds(bodyStmts: Seq[Stmt], predicatesToRemove: Seq[String]): Seq[Stmt] = {
+    def filterUnfolds(seqn: Seqn): Seqn = {
+      val seqnNoUnfolds = seqn.ss.filter {
+        case Unfold(pred) => !predicatesToRemove.contains(pred.loc.predicateName)
+        case _ => true
+      }
+      seqn.copy(ss = seqnNoUnfolds, scopedDecls = seqn.scopedDecls)(
+        pos = seqn.pos,
+        info = seqn.info,
+        errT = seqn.errT
+      )
+    }
+    bodyStmts.map {
+      case s: Seqn => filterUnfolds(s)
+      case other => other
+    }
+  }
+
+  private[this] def removeFolds(bodyStmts: Seq[Stmt], predicatesToRemove: Seq[String]): Seq[Stmt] = {
```

Comment:
Truly disgusting code duplication going on, will fix later once i can think of a nice way to do this. That said, I enjoy that we're doing fold/unfold removal in separate passes. We can easily consolidate fold/unfold removal in one pass but that violates SoC and makes me sad.